v2/src/logger.c

```diff
@@ -0,0 +1,35 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include <stdint.h>
+#include <stdarg.h>
+#include <stdio.h>
+
+#include "c_logging/logger.h"
+#include "c_logging/log_sink.h"
+#include "c_logging/log_sink_console.h"
+#include "c_logging/log_sink_etw.h"
+#include "c_logging/log_context.h"
+
+MU_DEFINE_ENUM_STRINGS_WITHOUT_INVALID(LOG_LEVEL, LOG_LEVEL_VALUES);
+
+const LOG_SINK* log_sinks[] = { &console_log_sink, &etw_log_sink, NULL };
+
+void logger_log(LOG_LEVEL log_level, LOG_CONTEXT_HANDLE log_context, const char* format, ...)
+{
+    uint32_t i;
+    va_list args;
+    va_start(args, format);
+
+    char message[LOG_MAX_MESSAGE_LENGTH];
+
+    vsprintf(message, format, args);
+
+    i = 0;
+    while (log_sinks[i] != NULL)
+    {
+        log_sinks[i++]->log_sink_log(log_level, log_context, message, __FILE__, __FUNCTION__, __LINE__);
```

Comment:
>__FUNCTION__ [](http://example.com/codeflow?start=80&length=12)

isn't __FUNCTION__ going to alway ebe "logger_log"? same for __LINE__, it will always be "31"

Reply:
HAHA, I knew there was a reason I had at some point a macro as entry point! Now I remembered

Reply:
But me has an idea. Stay tuned

Reply:
Latest version has a macro for this to pick up the function, file and line no.

Reply:
I initially thought of modeling these as properties in a special context, but likely that'd be overthinking it.